File: src/DeviceOptimizerHypercube.cpp
// this file is covered by the  GNU LESSER GENERAL PUBLIC LICENSE Version 3 or later
// please see LICENSE.txt for more details and licensing issues
// copyright Etienne de Foras ( the author )  mailto: etienne.deforas@gmail.com

#include "DeviceOptimizerHypercube.h"

#include <cassert>
////////////////////////////////////////////////////////////////////////////////
DeviceOptimizerHypercube::DeviceOptimizerHypercube():
    DeviceOptimizer()
{ }

//////////////////////////////////////////////////////////////////////////////
DeviceOptimizerHypercube::~DeviceOptimizerHypercube()
{ }
//////////////////////////////////////////////////////////////////////////////
OptimizerResult DeviceOptimizerHypercube::optimize()
{
    // divide hypercube into regular domain
    assert(_pDevice!=0);

    double dMeritOrig=compute_demerit();
    OpticalDevice deviceOrig(*_pDevice);

    if(_parameters.empty())
        return eNothingToOptimize;

    ParameterSet paramBest=_parameters;

    // compute nb of iterations
    int iNbIter=1;
    for(unsigned int i=0;i<paramBest.size();i++)
        iNbIter*=DIVIDE_ITER_BY_AXIS;

    double dBestMerit=SPOT_SIZE_INFINITY; // TODO init with actual solution

    bool bExit=false;
    while(!bExit)
    {
        ParameterSet paramBestIter=paramBest;
		vector<int> vDivide(paramBest.size(), 0);

        for(int i=0;i<iNbIter;i++)
        {
            ParameterSet paramBestTest=paramBest;
            for(unsigned int iP=0;iP<paramBestTest.size();iP++)
            {
                DeviceOptimizerParameter& dop=paramBestTest[iP];

                dop.dValue=dop.dMin+(dop.dMax-dop.dMin)*vDivide[iP]/ DIVIDE_ITER_BY_AXIS;
                assert(dop.dValue<=dop.dMax);
                assert(dop.dValue>=dop.dMin);
            }

            apply_parameter(paramBestTest);
            double dMerit=compute_demerit();

            if(dMerit<dBestMerit)
            {
                paramBestIter=paramBestTest;
                dBestMerit=dMerit;
            }
        
			//change next parameter in vDivide
			for (int i = 0; i < vDivide.size();i++)
			{
				if (vDivide[i] != DIVIDE_ITER_BY_AXIS)
				{
					vDivide[i]++;
					break;
				}
				else
					vDivide[i] = 0; // and jump next parameter
			}


		}

        paramBest=paramBestIter;

        // scale around best solution , divide by 2 each dimension test if box is too small then exit
		bExit = true;
        for(unsigned int iP=0;iP<paramBest.size();iP++)
        {
            DeviceOptimizerParameter& dop=paramBest[iP];
            double dCenter=dop.dValue;
            double dRadius=(dop.dMax-dop.dMin)/4.;

            dop.dMin=dCenter-dRadius;
            dop.dMax=dCenter+dRadius;

            // check domain exit
			if ((dop.dMax - dop.dMin) > dop.dResolution)
				bExit= false;
        }
    }

    if(dBestMerit<SPOT_SIZE_INFINITY/2)
    {
        if(dBestMerit<dMeritOrig)
        {
            _parameters=paramBest;
            apply_parameter(paramBest);
            return eBetterSolutionFound;
        }
    }

    //restore device original settings
    *_pDevice=deviceOrig;
    return eNoBetterSolution;
}
//////////////////////////////////////////////////////////////////////////////

File: python/DeviceOptimizerHyperCube.py
import numpy as np
import DeviceOptimizer
import OpticalDevice

DIVIDE_ITER_BY_AXIS=5

class DeviceOptimizerHypercube(DeviceOptimizer):
    def __init__(self):
        super().__init__()

    def __del__(self):
        pass

    def optimize(self):
        # divide hypercube into regular domain
        assert self._pDevice is not None

        dMeritOrig = self.compute_demerit()
        deviceOrig = OpticalDevice(self._pDevice)

        if not self._parameters:
            return DeviceOptimizer.eNothingToOptimize

        paramBest = self._parameters.copy()

        # compute number of iterations
        iNbIter = 1
        for _ in paramBest:
            iNbIter *= DIVIDE_ITER_BY_AXIS

        dBestMerit = DeviceOptimizer.SPOT_SIZE_INFINITY  # TODO init with actual solution

        bExit = False
        while not bExit:
            paramBestIter = paramBest.copy()
            vDivide = [0] * len(paramBest)

            for i in range(iNbIter):
                paramBestTest = paramBest.copy()
                for iP in range(len(paramBestTest)):
                    dop = paramBestTest[iP]

                    dop.dValue = dop.dMin + (dop.dMax - dop.dMin) * vDivide[iP] / DIVIDE_ITER_BY_AXIS
                    assert dop.dValue <= dop.dMax
                    assert dop.dValue >= dop.dMin

                self.apply_parameter(paramBestTest)
                dMerit = self.compute_demerit()

                if dMerit < dBestMerit:
                    paramBestIter = paramBestTest
                    dBestMerit = dMerit

                # change next parameter in vDivide
                for j in range(len(vDivide)):
                    if vDivide[j] != DIVIDE_ITER_BY_AXIS:
                        vDivide[j] += 1
                        break
                    else:
                        vDivide[j] = 0  # and jump to next parameter

            paramBest = paramBestIter

            # scale around best solution, divide by 2 each dimension test if box is too small then exit
            bExit = True
            for iP in range(len(paramBest)):
                dop = paramBest[iP]
                dCenter = dop.dValue
                dRadius = (dop.dMax - dop.dMin) / 4.0

                dop.dMin = dCenter - dRadius
                dop.dMax = dCenter + dRadius

                # check domain exit
                if (dop.dMax - dop.dMin) > dop.dResolution:
                    bExit = False

        if dBestMerit < DeviceOptimizer.SPOT_SIZE_INFINITY / 2:
            if dBestMerit < dMeritOrig:
                self._parameters = paramBest
                self.apply_parameter(paramBest)
                return DeviceOptimizer.eBetterSolutionFound

        # restore device original settings
        self._pDevice = deviceOrig
        return DeviceOptimizer.eNoBetterSolution